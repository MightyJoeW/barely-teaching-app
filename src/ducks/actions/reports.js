import uuid from "uuid";
import database from "../../firebase/firebase";

// ADD_REPORT
export const addReport = report => ({
  type: "ADD_REPORT",
  report
});

export const startAddReport = (reportData = {}) => {
  return dispatch => {
    const {
      student_name = "",
      note = "",
      period = 0,
      createdAt = 0
    } = reportData;
    const report = { student_name, note, period, createdAt };

    return database
      .ref("reports")
      .push(report)
      .then(ref => {
        dispatch(
          addReport({
            id: ref.key,
            ...report
          })
        );
      });
  };
};

// REMOVE_REPORT
export const removeReport = ({ id } = {}) => ({
  type: "REMOVE_REPORT",
  id
});

export const startRemoveReport = ({ id } = {}) => {
  return dispatch => {
    return database
      .ref(`reports/${id}`)
      .remove()
      .then(() => {
        dispatch(removeReport({ id }));
      });
  };
};

// EDIT_REPORT
export const editReport = (id, updates) => ({
  type: "EDIT_REPORT",
  id,
  updates
});

export const startEditReport = (id, updates) => {
  return dispatch => {
    return database
      .ref(`reports/${id}`)
      .update(updates)
      .then(() => {
        dispatch(editReport(id, updates));
      });
  };
};

// SET_REPORTS
export const setReports = reports => ({
  type: "SET_REPORTS",
  reports
});

export const startSetReports = () => {
  return dispatch => {
    return database
      .ref("reports")
      .once("value")
      .then(snapshot => {
        const reports = [];

        snapshot.forEach(childSnapshot => {
          reports.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          });
        });

        dispatch(setReports(reports));
      });
  };
};
