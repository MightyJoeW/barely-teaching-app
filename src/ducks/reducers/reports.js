// Reports Reducer

const reportsReducerDefaultState = [];

export default (state = reportsReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_REPORT":
      return [...state, action.report];
    case "REMOVE_REPORT":
      return state.filter(({ id }) => id !== action.id);
    case "EDIT_REPORT":
      return state.map(report => {
        if (report.id === action.id) {
          return { ...report, ...action.updates };
        } else {
          return report;
        }
      });
    default:
      return state;
  }
};
