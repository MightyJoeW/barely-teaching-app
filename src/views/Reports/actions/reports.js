import uuid from "uuid";

// ADD_REPORT
export const addReport = (
  { student_name = "", note = "", period = 0, createdAt = 0 } = {}
) => ({
  type: "ADD_REPORT",
  report: {
    id: uuid(),
    student_name,
    note,
    period,
    createdAt
  }
});

// REMOVE_REPORT
export const removeReport = ({ id } = {}) => ({
  type: "REMOVE_REPORT",
  id
});

// EDIT_REPORT
export const editReport = (id, updates) => ({
  type: "EDIT_REPORT",
  id,
  updates
});
