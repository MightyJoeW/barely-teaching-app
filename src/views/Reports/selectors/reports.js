import moment from "moment";

// Get visible reports
export default (reports, { text, sortBy, startDate, endDate }) => {
  return reports
    .filter(report => {
      const createdAtMoment = moment(report.createdAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, "day")
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, "day")
        : true;
      const textMatch = report.student_name
        .toLowerCase()
        .includes(text.toLowerCase());

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return a.createdAt < b.createdAt ? 1 : -1;
      } else if (sortBy === "period") {
        return a.period < b.period ? 1 : -1;
      }
    });
};
