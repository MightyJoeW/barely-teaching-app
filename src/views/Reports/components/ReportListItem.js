import React from "react";
import { connect } from "react-redux";
import { removeReport } from "../actions/reports";
const ReportListItem = ({ dispatch, id, student_name, period, createdAt }) => (
  <div>
    <h3>{student_name} </h3>
    <p>
      {period} - {createdAt}
    </p>
    <button
      onClick={() => {
        dispatch(removeReport({ id }));
      }}
    >
      Remove
    </button>
  </div>
);

export default connect()(ReportListItem);
