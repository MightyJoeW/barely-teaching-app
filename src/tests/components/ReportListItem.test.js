import React from "react";
import { shallow } from "enzyme";
import reports from "../fixtures/reports";
import ReportListItem from "../../views/Reports/components/ReportListItem";

test("should render ReportListItem correctly", () => {
  const wrapper = shallow(<ReportListItem {...reports[0]} />);
  expect(wrapper).toMatchSnapshot();
});
