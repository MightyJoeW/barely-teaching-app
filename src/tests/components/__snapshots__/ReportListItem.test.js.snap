// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ReportListItem correctly 1`] = `
<Link
  className="list-item"
  replace={false}
  to="/reports/edit/1"
>
  <div>
    <h3
      className="list-item__title"
    >
      Jack
       
    </h3>
    <span
      className="list-item__subtitle"
    >
      December 31st, 1969
    </span>
  </div>
  <h3
    className="list-item__data"
  >
     
    1st
     
  </h3>
</Link>
`;
